src/main/java/net/openhft/chronicle/wire/Wires.java

```diff
@@ -674,34 +676,63 @@ public static WireOut writeDate(Date date, ValueOut out) {
         }
 
         public static Date parseDate(ValueIn in) {
-            String text = in.text().trim();
-            try {
-                return new Date(Long.parseLong(text));
-            } catch (NumberFormatException nfe) {
+            final String text = in.text().trim();
+            if (text.length() < 1) {
+                throw new IORuntimeException("At least one character (e.g. '0') must be present in order to deserialize a Date object");
+            }
+            final char firstChar = text.charAt(0);
+
+            // Check if it is a number
+            if (firstChar == '+' || firstChar == '-' || Character.isDigit(firstChar)) {
+                boolean isAllNum = true;
+                for (int i = 1; i < text.length(); i++) {
+                    if (!Character.isDigit(text.charAt(i))) {
+                        isAllNum = false;
+                        break;
+                    }
+                }
+                if (isAllNum) {
+                    try {
+                        // It is a number and not any of the other formats as they all contain non digits
+                        return new Date(Long.parseLong(text));
+                    } catch (NumberFormatException nfe) {
+                        throw new IORuntimeException(nfe);
+                    }
+                }
+            }
+
+            // Check if it is a string of format "yyyy-mm-dd"
+            if (text.length() == SDF_4_STRING.length()) {
                 try {
-                    // Make sure to always synchronize on these fields in exactly this order or else there might be deadlocks
-                    synchronized (SDF_2) {
-                        return SDF_2.parse(text);
+                    synchronized (SDF_4) {
+                        // Since we ruled out it is a number and SDF_$ is the only one with this length it must be SDF_4
+                        return SDF_4.parse(text);
                     }
                 } catch (ParseException pe) {
+                    throw new IORuntimeException(pe);
+                }
+            }
+
+            // Try the other remaining formats
+            // Todo: optimize away exception chaining
+            try {
+                // Make sure to always synchronize on these fields in exactly the order below if used elsewhere or otherwise there might be deadlocks.
+                // SDF_2 is the most likely one since it is used for serialization in writeDate()
```

Comment:
can't we just eliminate the possibility of deadlocks if we remove the nesting?

or at least the nesting of the synchronized blocks? e.g.

```
try {
   synchronized(thing) {
      return thing.parse(text);
   }
} catch (ParseException ex) {
  // ignore
}

try {
   synchronized(thing2) {
      thing2.parse(text);
   }
} catch (ParseException ex) {
  // ignore
}

// none parsed, throw
throw new IORuntimeException(...)


```

Reply:
I wonder how I did not think about this in the first place...